### PR TITLE
Update order if there is an error while Stripe confirms payment

### LINKED ISSUE #1740: UPE - Errors at checkout are not reflected on order status
**Describe the bug**

This is a mirror of the same WCPay issue at Automattic/woocommerce-payments#2702.

> When completing a checkout using credit/debit cards with the UPE, if the payment information provided causes an error, we do not update the order with this event. The previous, parent payment gateway would update the order's status to `Failed` at this stage and a note would be left on the order explaining the specific error message. With the UPE, in this case, we simply display the error at checkout to the user, without sending any request to update the order. This could be confusing to a site owner, since the order's status remains as `Pending payment` and is not associated with other failed orders.

**To Reproduce**
1. Enable the UPE following the steps in #1659. 
2. Add an item to your cart and proceed to checkout and wait for the UPE to load.
3. Select credit/debit card within the UPE and use the details from one of [these test cards](https://stripe.com/docs/testing#cards-responses) to effect a failing response. Note that an error is displayed at checkout. 
4. From your site's admin section, view your latest orders and find the corresponding order for your checkout. Note that the status is still `Pending payment` with no notes added to the order.
5. You can repeat these steps with the UPE disabled to see how the order should be reflecting the error instead. The order should be updated to `Failed` and there should be notes on the order explaining the error.

**Expected behavior**

Order status should change to `Failed`. Order should contain note explaining transaction error.

**Screenshots**

**Environment (please complete the following information):**
 - WordPress Version
 - WooCommerce Version
 - Stripe Plugin Version
 - Browser [e.g. chrome, safari] and Version
 - Any other plugins installed

**Additional context**

### PR DESCRIPTION
**Pull request by FangedParakeet**
# Changes proposed in this Pull Request:

Issue: Link to the GitHub issue this PR addresses (if appropriate).
Fixes #1740 

Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.

This PR adds an AJAX function to handle errors that may occur while Stripe is confirming payment for an order. Previously we simply threw an error on the client and displayed a message to the user if `confirmPayment` or `confirmSetup` returned an error: however, we should be updating the order appropriately as well, in this scenario. This is predominantly catered towards payment methods without redirection (e.g. cards, SEPA); redirection payment methods (e.g. giropay) succeed or fail after authorisation and we handle these errors safely after the redirect. Until now we did not really have any error handling if an error appeared prior to this.

Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.

# Testing instructions

Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?

1. Please pull the latest changes from this PR and enable the UPE, of course.
2. Add an item to cart and complete a checkout using one of [these cards](https://stripe.com/docs/testing#cards-responses) that will elicit an error response.
3. Notice that an error is produced at checkout as usual. 
4. Go to your Orders inside WP-Admin and notice that an order has been created and that the status of the order is now Failed. 
5. Complete the order or continue producing errors and confirming that the status of the order is updated appropriately.
6. Test the above steps from the Pay for Order page and confirm the same results. 

-------------------
- [ ] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

(remove me)
- [ ] Did I add a title? A descriptive, yet concise, title.
- [ ] How can this code break?
- [ ] What are we doing to make sure this code doesn't break?

### CODE REVIEW

File: client/api/index.js [hunk 1]
```diff
@@ -379,4 +379,23 @@ export default class WCStripeAPI {
...
+		this.request( getAjaxUrl( 'update_failed_order' ), {
+			intent_id: intentId,
+			order_id: orderId,
+			_ajax_nonce: getStripeServerData()?.updateFailedOrderNonce,
+		} )
+			.catch( ( ) => {
+				// If something goes wrong here,
+				// we would still rather throw the Stripe error rather than this one.
+				return;
+			} );
```
Comment (FangedParakeet) [inline client/api/index.js:399]: This is a bit ugly, but essentially if this AJAX request errors out for any reason--which it really shouldn't--I would like to throw the error produced by `confirmPayment`/`confirmSetup` instead of throwing any errors from this request. Additionally, there is no useful data returned by this AJAX request, so we basically just want to swallow the output here whatever happens.

File: includes/class-wc-stripe-intent-controller.php [hunk 2]
```diff
@@ -588,6 +590,61 @@ public function update_order_status_ajax() {
...
+				}
+
+				// Use the last charge within the intent to proceed.
+				if ( isset( $intent->charges ) ) {
+					$charge = end( $intent->charges->data );
+					$this->process_response( $charge, $order );
+				} else {
+					// TODO: Add implementation for setup intents.
+					$this->process_response( $intent, $order );
+				}
```
Comment (FangedParakeet) [inline includes/class-wc-stripe-intent-controller.php]: In most cases, we should discover an error on the intent inside `last_payment_error`, but just in case the error is clean somehow, but there is an error or something else unusual on the charge, `process_response` can handle this case: if the status on the charge is `failed`, `process_response` will throw an exception and we will fail the order.

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 3]
```diff
@@ -448,6 +449,7 @@ public function process_payment( $order_id ) {
 		return [
 			'result'         => 'success',
 			'payment_needed' => $payment_needed,
+			'order_id'       => $order_id,
```
Comment (FangedParakeet) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:459]: We are returning the order ID inside the `redirect_url` parameter returned here, so I considered pulling this from the URL on the client instead of including it twice, so we could send it back if there was an error, but in the end this just seemed simpler than adding additional code to parse the URL in Javascript.

File: client/classic/upe/index.js [hunk 4]
```diff
@@ -468,6 +468,7 @@ jQuery( function ( $ ) {
 				},
 			} );
 			if ( error ) {
+				await api.updateFailedOrder( paymentIntentId, response.order_id );
```
Comment (bborman22) [inline client/classic/upe/index.js]: When testing on the Pay for Order page, got a response is not defined error. I think you can just use the `orderId` variable from above here though and it works as expected then.
```suggestion
				await api.updateFailedOrder( paymentIntentId, orderId );
```

File: includes/class-wc-stripe-intent-controller.php [hunk 5]
```diff
@@ -604,6 +606,62 @@ public function update_order_status_ajax() {
...
+				$error = $intent->last_payment_error;
+
+				if ( ! empty( $error ) ) {
+					WC_Stripe_Logger::log( 'Error when processing payment: ' . $error->message );
+					throw new WC_Stripe_Exception( __( "We're not able to process this payment. Please try again later.", 'woocommerce-gateway-stripe' ) );
+				}
+
+				// Use the last charge within the intent to proceed.
+				$gateway = $this->get_gateway();
+				if ( isset( $intent->charges ) ) {
```
Comment (bborman22) [inline includes/class-wc-stripe-intent-controller.php]: This fallback area is a little awkward just because I'm not even sure how it would be reached in a real world scenario, but I like that it is at least there! To test this area, I just made the failed order call before even attempting to confirm the intent. This led to the creation of an order on-hold, but it didn't have any charge or payment intent IDs attached to allow the merchant to manually process the charge if desired. What I found was that we should also be checking that `$intent->charges->data` is not empty before using the charge in the `process_response` call instead of the intent.

Long story short, I think the conditional should have: 
```suggestion
				if ( isset( $intent->charges ) && ! empty( $intent->charges->data ) ) {
```

File: client/api/index.js (hunk 1 above)
Comment (bborman22) [reply client/api/index.js:399]: This makes sense to me to just avoid an unnecessary error.

**Review by bborman22 — CHANGES_REQUESTED** [review]
Overall this worked great and was fun to test with a "stolen card" 😆 . I just had two notes to address: one to fix the pay for order page and the other to make the fallback process provide the merchant with a little more information.

Once those two are addressed, I think this should be good to go!

File: client/classic/upe/index.js (hunk 4 above)
Comment (FangedParakeet) [reply client/classic/upe/index.js]: D'oh! Believe it or not, but I specifically designed this function with these separate parameters in mind, so that I'd be able to use it with `orderId` in `handleUPEOrderPay` as well...and then I forgot about it and just copy-pasted away. Nice catch, cheers!

File: includes/class-wc-stripe-intent-controller.php (hunk 5 above)
Comment (FangedParakeet) [reply includes/class-wc-stripe-intent-controller.php]: Cool, suggestion seems sound: committed. As stated, I'm also not sure how to get to that particular section of code without any wizardry, but you know what they say about death, taxes, and Stripe doing strange things unexpectedly...

**Comment by FangedParakeet** [comment]
Cheers for the quick review, @bborman22! Just applied both your suggestions: if you wouldn't mind giving this one last little once over to let me know if it's ready for the floor.

**Review by bborman22 — APPROVED** [review]
Thanks for those fixes everything worked as expected!

LGTM!
